File: jar/src/main/kotlin/jline/lang/processes/Marked.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.processes;

import jline.util.matrix.Matrix;

import java.io.Serializable;

/**
 * An abstract class for marked point processes
 */

abstract public class Marked extends MarkovModulated implements Serializable {
    // A Markov-modulated Markovian point process, possibly non-renewal

    public Marked(String name, int numParam) {
        super(name, numParam);
    }

    public Matrix D(int i) {
        return this.process.get(i);
    }

    public Matrix D(int i, int k) {
        if (i == 1) {
            return this.process.get(2 + k);
        } else {
            return null;
        }
    }

    public Matrix getD1k(int k) {
        return (Matrix) this.getParam(2 + k).getValue();
    }

}

File: matlab/src/lang/processes/Marked.m
classdef Marked < MarkovModulated
    % An abstract class for Markov-modulated marked processes
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    methods (Hidden)
        %Constructor
        function self = MarkedMarkovModulated(name, numParam)
            % SELF = MARKOVMODULATED(NAME, NUMPARAM)

            self@PointProcess(name, numParam);
        end
    end

    methods
        function X = sample(self, n)
            % X = SAMPLE(N)
            line_error(mfilename,'An abstract method was called. The function needs to be overridden by a subclass.');
        end

        function P = getEmbedded(self)
            % P = GETEMBEDDED()
            %
            % Get DTMC embedded at event arrival times

            P = map_embedded(self.getProcess);
        end

        function pie = getEmbeddedProb(self)
            % PIE = GETEMBEDDEDRPOB()
            %
            % Solve DTMC embedded embedded at event arrival times

            pie = map_pie(self.getProcess);
        end
    end

    methods (Abstract)
        phases = getNumberOfPhases(self);
    end

end

